packages/@lwc/engine-core/src/framework/base-bridge-element.ts

```diff
@@ -125,6 +132,43 @@ export function HTMLBridgeElementFactory(
         };
     }
     defineProperties(HTMLBridgeElement.prototype, descriptors);
+    // observable attributes
+    HTMLBridgeElement.prototype.attributeChangedCallback = function (
+        this: HTMLElement,
+        attrName: string,
+        oldValue: string,
+        newValue: string
+    ) {
+        if (oldValue === newValue) {
+            // Ignore same values.
+            return;
+        }
+        const propName = attributeToPropMap[attrName];
+        if (isUndefined(propName)) {
+            if (!isUndefined(superAttributeChangedCallback)) {
+                // delegate unknown attributes to the super.
+                superAttributeChangedCallback.apply(this, arguments);
+            }
+            return;
+        }
+        if (!isAttributeLocked(this, attrName)) {
+            // Ignore changes triggered by the engine itself during:
+            // * diffing when public props are attempting to reflect to the DOM
+            // * component via `this.setAttribute()`, should never update the prop
+            // Both cases, the setAttribute call is always wrapped by the unlocking of the
+            // attribute to be changed
+            return;
+        }
+        // Reflect attribute change to the corresponding property when changed from outside.
+        (this as any)[propName] = newValue;
+    };
+    // Specify attributes for which we want to reflect changes back to their corresponding
+    // properties via attributeChangedCallback.
+    defineProperty(HTMLBridgeElement, 'observedAttributes', {
+        get() {
+            return [...((SuperClass as any).observedAttributes || []), ...keys(attributeToPropMap)];
```

Comment:
Instead of adding an `any` type guard you can extend `HTMLElementConstructor` with an optional `observedAttributes` property.

```suggestion
            return [...(SuperClass.observedAttributes || []), ...keys(attributeToPropMap)];
```

Reply:
when I do this, I get a ts error saying that superclass doesn't have observedAttributes, even though it is marked as HTMLElementConstructor

Reply:
The following types work for me on your PR.

```ts
type AttributeChangedCallback = (
    this: HTMLElement,
    attrName: string,
    oldValue: string,
    newValue: string
) => void;

interface CustomElement extends HTMLElement {
    attributeChangedCallback?: AttributeChangedCallback;
}

interface HTMLElementConstructor {
    prototype: CustomElement;
    new (): CustomElement;

    observedAttributes?: string[]
}
```